TCP lwIP Error Connecting Sock Problem on ESP32

@gschorcht  @PeterKietzmann 
I switch to lwIP and modified the offical TCP example the code is here.  I run ESP32 802.11 board.

I have TCP Server in Java, this is a TCP client.  I also modified esp_wifi_params.h to my AP credentials just in case. I am sure that ipv6 is valid because udp example runs OK using same ipv6 address. However, Error is given in 
`if (sock_tcp_connect(&sock, &remote, 0, 0) < 0) {
        puts("Error connecting sock");
        return 1;
    }` 

Here. Also it connects to my Access point i am sure that because i am able to ping the ESP32.

What it could be the reason behind this?



```C
#include "net/af.h"
#include "net/ipv6/addr.h"
#include "net/sock/tcp.h"
#include "xtimer.h"
uint8_t buf[128];

/* import "ifconfig" shell command, used for printing addresses */
extern int _gnrc_netif_config(int argc, char **argv);

int main(void)
{

	/* print network addresses */
    puts("Configured network interfaces2:");
    _gnrc_netif_config(0, NULL);
	
    xtimer_sleep(5);
    sock_tcp_t sock;	
    printf("****TCP CLIENT*3*3*3*");
    int res;
    sock_tcp_ep_t remote = SOCK_IPV6_EP_ANY;
    remote.port = 12345;
    ipv6_addr_from_str((ipv6_addr_t *)&remote.addr,
                       "fe80::4c2c:2f99:9ae8:73cb");
    if (sock_tcp_connect(&sock, &remote, 0, 0) < 0) {
        puts("Error connecting sock");
        return 1;
    }
    puts("Sending \"Hello!\"");
    if ((res = sock_tcp_write(&sock, "Hello!", sizeof("Hello!"))) < 0) {
        puts("Errored on write");
    }
    else {
        if ((res = sock_tcp_read(&sock, &buf, sizeof(buf),
                                 SOCK_NO_TIMEOUT)) < 0) {
            puts("Disconnected");
        }
        printf("Read: \"");
        for (int i = 0; i < res; i++) {
            printf("%c", buf[i]);
        }
        puts("\"");
    }
    sock_tcp_disconnect(&sock);
    return res;
}
```


I have added necessary modules also, 

```Makefile
# name of your application
APPLICATION = tcp-simple-server
# If no BOARD is found in the environment, use this default:
BOARD ?= esp32-wroom-32

# This has to be the absolute path to the RIOT base directory:
RIOTBASE ?= $(CURDIR)/../..

BOARD_INSUFFICIENT_MEMORY := arduino-duemilanove arduino-leonardo \
                             arduino-mega2560 arduino-nano \
                             arduino-uno blackpill bluepill calliope-mini \
                             chronos hifive1 i-nucleo-lrwan1 mega-xplained \
                             microbit msb-430 msb-430h \
                             nucleo-f031k6 nucleo-f042k6 nucleo-f303k8 \
                             nucleo-l031k6 nucleo-f030r8 nucleo-f070rb \
                             nucleo-f072rb nucleo-f103rb nucleo-f302r8 \
                             nucleo-f334r8 nucleo-l053r8 saml10-xpro \
                             saml11-xpro spark-core stm32f0discovery \
                             stm32l0538-disco telosb \
                             waspmote-pro wsn430-v1_3b wsn430-v1_4 z1

# Add also the shell, some shell commands
USEMODULE += auto_init_gnrc_netif 
USEMODULE += gnrc_netdev_default


USEMODULE += gnrc_icmpv6_echo

USEMODULE += lwip lwip_ipv6_autoconfig lwip_sock_ip lwip_netdev
USEMODULE += lwip_udp lwip_sock_udp
USEMODULE += lwip_tcp lwip_sock_tcp
USEMODULE += shell
USEMODULE += shell_commands
USEMODULE += ps
USEMODULE += netstats_l2
USEMODULE += netstats_ipv6
USEMODULE += netstats_rpl

USEMODULE += ipv6_addr
USEMODULE += od
USEMODULE += netdev_default
USEMODULE += esp_wifi 
DEVELHELP ?= 1
# Change this to 0 show compiler invocation lines by default:
QUIET ?= 1

include $(RIOTBASE)/Makefile.include

# Set a custom channel if needed
ifneq (,$(filter cc110x,$(USEMODULE)))          # radio is cc110x sub-GHz
  DEFAULT_CHANNEL ?= 0
  CFLAGS += -DCC110X_DEFAULT_CHANNEL=$(DEFAULT_CHANNEL)
else
  ifneq (,$(filter at86rf212b,$(USEMODULE)))    # radio is IEEE 802.15.4 sub-GHz
    DEFAULT_CHANNEL ?= 5
    CFLAGS += -DIEEE802154_DEFAULT_SUBGHZ_CHANNEL=$(DEFAULT_CHANNEL)
  else                                          # radio is IEEE 802.15.4 2.4 GHz
    DEFAULT_CHANNEL ?= 26
    CFLAGS += -DIEEE802154_DEFAULT_CHANNEL=$(DEFAULT_CHANNEL)
  endif
endif
```

> By the way these are the warnings when i compile the program
> ```
> /data/riotbuild/RIOT-master/examples/tcp-simple-server/bin/esp32-wroom-32/lwip_api.a(tcpip.o): In function `tcpip_try_callback':
> tcpip.c:(.text+0x32c): warning: undefined reference to `sys_mbox_trypost_fromisr'
> /data/riotbuild/RIOT-master/examples/tcp-simple-server/bin/esp32-wroom-32/lwip_api.a(tcpip.o): In function `tcpip_send_msg_wait_sem':
> tcpip.c:(.text+0x35e): warning: undefined reference to `sys_mbox_trypost_fromisr
> ```

BTW, I was trying investige these undefined symbols a year ago, but I don't remeber what the problem was.

> 
> 
> > By the way these are the warnings when i compile the program
> > ```
> > /data/riotbuild/RIOT-master/examples/tcp-simple-server/bin/esp32-wroom-32/lwip_api.a(tcpip.o): In function `tcpip_try_callback':
> > tcpip.c:(.text+0x32c): warning: undefined reference to `sys_mbox_trypost_fromisr'
> > /data/riotbuild/RIOT-master/examples/tcp-simple-server/bin/esp32-wroom-32/lwip_api.a(tcpip.o): In function `tcpip_send_msg_wait_sem':
> > tcpip.c:(.text+0x35e): warning: undefined reference to `sys_mbox_trypost_fromisr
> > ```
> 
> BTW, I was trying to investige these undefined symbols a year ago, but I don't remeber what the problem was.

Thanks for your reply. In conclusion, I couldn't manage to run a TCP application using neither lwip nor gnrc in ESP32. As you mentioned, will ESP8266 able to run a TCP application? I am trying to catch up the TCP application to my thesis and conferance paper.

@gschorcht  I tried to build for ESP32 however i got this error message during compile process.

`/data/riotbuild/pkg/lwip/contrib/lwip.c: In function 'lwip_bootstrap':
/data/riotbuild/pkg/lwip/contrib/lwip.c:137:30: error: passing argument 2 of 'netif_add' from incompatible pointer type [-Werror=incompatible-pointer-types]
     if (netif_add(&netif[0], &_esp_wifi_dev, lwip_netdev_init,
                              ^
In file included from /data/riotbuild/examples/tcp-simple-client/bin/pkg/esp32-wroom-32/lwip/src/include/lwip/tcpip.h:46:0,
                 from /data/riotbuild/pkg/lwip/contrib/lwip.c:16:
/data/riotbuild/examples/tcp-simple-client/bin/pkg/esp32-wroom-32/lwip/src/include/lwip/netif.h:416:15: note: expected 'const ip4_addr_t * {aka const struct ip4_addr *}' but argument is of type 'esp_wifi_netdev_t * {aka struct <anonymous> *}'
 struct netif *netif_add(struct netif *netif,
               ^
/data/riotbuild/pkg/lwip/contrib/lwip.c:137:46: error: passing argument 3 of 'netif_add' from incompatible pointer type [-Werror=incompatible-pointer-types]
     if (netif_add(&netif[0], &_esp_wifi_dev, lwip_netdev_init,
                                              ^
In file included from /data/riotbuild/examples/tcp-simple-client/bin/pkg/esp32-wroom-32/lwip/src/include/lwip/tcpip.h:46:0,
                 from /data/riotbuild/pkg/lwip/contrib/lwip.c:16:
/data/riotbuild/examples/tcp-simple-client/bin/pkg/esp32-wroom-32/lwip/src/include/lwip/netif.h:416:15: note: expected 'const ip4_addr_t * {aka const struct ip4_addr *}' but argument is of type 'err_t (*)(struct netif *) {aka signed char (*)(struct netif *)}'
 struct netif *netif_add(struct netif *netif,
               ^
/data/riotbuild/pkg/lwip/contrib/lwip.c:138:19: error: passing argument 4 of 'netif_add' from incompatible pointer type [-Werror=incompatible-pointer-types]
                   tcpip_input) == NULL) {
                   ^
In file included from /data/riotbuild/examples/tcp-simple-client/bin/pkg/esp32-wroom-32/lwip/src/include/lwip/tcpip.h:46:0,
                 from /data/riotbuild/pkg/lwip/contrib/lwip.c:16:
/data/riotbuild/examples/tcp-simple-client/bin/pkg/esp32-wroom-32/lwip/src/include/lwip/netif.h:416:15: note: expected 'const ip4_addr_t * {aka const struct ip4_addr *}' but argument is of type 'err_t (*)(struct pbuf *, struct netif *) {aka signed char (*)(struct pbuf *, struct netif *)}'
 struct netif *netif_add(struct netif *netif,
               ^
/data/riotbuild/pkg/lwip/contrib/lwip.c:137:9: error: too few arguments to function 'netif_add'
     if (netif_add(&netif[0], &_esp_wifi_dev, lwip_netdev_init,
         ^
In file included from /data/riotbuild/examples/tcp-simple-client/bin/pkg/esp32-wroom-32/lwip/src/include/lwip/tcpip.h:46:0,
                 from /data/riotbuild/pkg/lwip/contrib/lwip.c:16:
/data/riotbuild/examples/tcp-simple-client/bin/pkg/esp32-wroom-32/lwip/src/include/lwip/netif.h:416:15: note: declared here
 struct netif *netif_add(struct netif *netif,
               ^
cc1: all warnings being treated as errors
/data/riotbuild/Makefile.base:83: recipe for target '/data/riotbuild/examples/tcp-simple-client/bin/esp32-wroom-32/lwip_contrib/lwip.o' failed
make[2]: *** [/data/riotbuild/examples/tcp-simple-client/bin/esp32-wroom-32/lwip_contrib/lwip.o] Error 1
/data/riotbuild/Makefile.base:20: recipe for target 'ALL--/data/riotbuild/pkg/lwip/contrib' failed
make[1]: *** [ALL--/data/riotbuild/pkg/lwip/contrib] Error 2
/data/riotbuild/examples/tcp-simple-client/../../Makefile.include:473: recipe for target '/data/riotbuild/examples/tcp-simple-client/bin/esp32-wroom-32/application_tcp-simple-client.a' failed
make: *** [/data/riotbuild/examples/tcp-simple-client/bin/esp32-wroom-32/application_tcp-simple-client.a] Error 2
riotbuild@67661a4da1bf:~/examples/tcp-simple-client$ 
`

--- a/cpu/esp32/esp-wifi/esp_wifi_netdev.c
+++ b/cpu/esp32/esp-wifi/esp_wifi_netdev.c
@@ -61,7 +61,7 @@
  * not provide an argument that could be used as pointer to the ESP WiFi
  * device which triggers the interrupt.
  */
-static esp_wifi_netdev_t _esp_wifi_dev;
+esp_wifi_netdev_t _esp_wifi_dev;
 static const netdev_driver_t _esp_wifi_driver;
 
 /* device thread stack */
@@ -202,7 +202,7 @@ static wifi_config_t wifi_config_sta = {
     }
 };
 
-static void esp_wifi_setup (esp_wifi_netdev_t* dev)
+void esp_wifi_setup (esp_wifi_netdev_t* dev)
 {
     DEBUG("%s: %p\n", __func__, dev);
 

--- a/cpu/esp32/esp-wifi/esp_wifi_netdev.c
+++ b/cpu/esp32/esp-wifi/esp_wifi_netdev.c
@@ -408,7 +408,7 @@ static int _esp_wifi_get(netdev_t *netdev, netopt_t opt, void *val, size_t max_l
 
     switch (opt) {
         case NETOPT_ADDRESS:
-            assert(max_len == ETHERNET_ADDR_LEN);
+            assert(max_len >= ETHERNET_ADDR_LEN);
             esp_wifi_get_mac(ESP_MAC_WIFI_STA,(uint8_t *)val);
             return ETHERNET_ADDR_LEN;
         case NETOPT_IS_WIRED:

--- a/pkg/lwip/contrib/lwip.c
+++ b/pkg/lwip/contrib/lwip.c
@@ -38,6 +38,10 @@
 #include "socket_zep_params.h"
 #endif
 
+#if defined(MODULE_ESP_WIFI) && defined(CPU_ESP32)
+#include "esp-wifi/esp_wifi_netdev.h"
+#endif
+
 #include "lwip.h"
 
 #define ENABLE_DEBUG    (0)
@@ -59,6 +63,10 @@
 #define LWIP_NETIF_NUMOF        (sizeof(socket_zep_params) / sizeof(socket_zep_params[0]))
 #endif
 
+#if defined(MODULE_ESP_WIFI) && defined(CPU_ESP32)  /* is mutual exclusive with above ifdef */
+#define LWIP_NETIF_NUMOF        (1)
+#endif
+
 #ifdef LWIP_NETIF_NUMOF
 static struct netif netif[LWIP_NETIF_NUMOF];
 #endif
@@ -77,6 +85,11 @@ static mrf24j40_t mrf24j40_devs[LWIP_NETIF_NUMOF];
 
 #ifdef MODULE_SOCKET_ZEP
 static socket_zep_t socket_zep_devs[LWIP_NETIF_NUMOF];
+#endif
+
+#if defined(MODULE_ESP_WIFI) && defined(CPU_ESP32)
+extern esp_wifi_netdev_t _esp_wifi_dev;
+extern void esp_wifi_setup (esp_wifi_netdev_t* dev);
 #endif
 
 void lwip_bootstrap(void)
@@ -118,6 +131,13 @@ void lwip_bootstrap(void)
             DEBUG("Could not add socket_zep device\n");
             return;
         }
+    }
+#elif defined(MODULE_ESP_WIFI) && defined(CPU_ESP32)
+    esp_wifi_setup(&_esp_wifi_dev);
+    if (netif_add(&netif[0], &_esp_wifi_dev, lwip_netdev_init,
+                  tcpip_input) == NULL) {
+        DEBUG("Could not add esp_wifi device\n");
+        return;
     }
 #endif
     if (netif[0].state != NULL) {

--- a/pkg/lwip/include/lwipopts.h
+++ b/pkg/lwip/include/lwipopts.h
@@ -143,6 +143,15 @@ extern "C" {
 
 #define TCPIP_THREAD_STACKSIZE  (THREAD_STACKSIZE_DEFAULT)
 
+#if defined(CPU_ESP32) && !defined(DOXYGEN)
+/**
+ * In ESP32, the thread that is dealing with hardware interrupts of the WiFi
+ * interface has a priority of 1. This thread should have a higher priority
+ * than lwIP's TCP/IP thread.
+ */
+#define TCPIP_THREAD_PRIO       (2)
+#endif
+
 #define MEM_ALIGNMENT           (4)
 #ifndef MEM_SIZE
 /* packet buffer size of GNRC + stack for TCP/IP */

